main/src/ca/uwaterloo/flix/language/phase/PatternExhaustiveness.scala

```diff
@@ -0,0 +1,524 @@
+/*
+ * Copyright 2017 Jason Mittertreiner
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ca.uwaterloo.flix.language.phase
+
+import ca.uwaterloo.flix.api.Flix
+import ca.uwaterloo.flix.language.ast.Symbol.EnumSym
+import ca.uwaterloo.flix.language.ast.Type.{Apply, Arrow, Bool, FTuple, Native, Ref}
+import ca.uwaterloo.flix.language.ast.{Type, TypedAst}
+import ca.uwaterloo.flix.language.ast.TypedAst.Pattern
+import ca.uwaterloo.flix.language.ast.TypedAst.Pattern.Var
+import ca.uwaterloo.flix.language.errors.NonExhaustiveMatchError
+import ca.uwaterloo.flix.language.phase.PatternExhaustiveness.Exhaustiveness.{Exhaustive, NonExhaustive}
+import ca.uwaterloo.flix.language.phase.PatternExhaustiveness.TypeConstructor.{BigInt, Char, Enum, False, Float32, Float64, Int16, Int32, Int64, Int8, Str, True, Tuple, Unit, Wild}
+import ca.uwaterloo.flix.language.{CompilationError, GenSym}
+import ca.uwaterloo.flix.util.Validation
+import ca.uwaterloo.flix.util.Validation._
+
+/**
+  * The Pattern Exhaustiveness phase checks pattern matches for exhaustiveness
+  *
+  * A pattern match is exhaustive if there exists a type correct expression which is not matched by any of the patterns
+  * in the pattern match
+  *
+  * Implemented as described in
+  * MARANGET, L. (2007). Warnings for pattern matching. Journal of Functional Programming, 17(3), 387-421. doi:10.1017/S0956796807006223
+  *
+  * The algorithm works by recursively "peeling off layers" of the pattern match. It check if the outermost type
+  * constructors are exhaustive, then if so, recursively calls itself on the arguments. If it finds a pattern which
+  * is not exhaustive during a recursive case, it uses it to built up a pattern that is not matched by the full
+  * pattern match and returns the result.
+  *
+  */
+object PatternExhaustiveness extends Phase[TypedAst.Root, TypedAst.Root] {
+
+  sealed trait TypeConstructor
+  object TypeConstructor {
+    case class True() extends TypeConstructor
```

Comment:
In Scala singleton case classes are case objects, i.e.

```
case object True extends TypeConstructor
```

instead of 

```
case class True() extends TypeConstructor
```

In Flix we don't have this mess :)